### ISSUE TITLE
Performance issue with ioQueue

### ISSUE THREAD

**Issue by Madimo** [SDWebImage/SDWebImage#3456]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | iOS
 Platform Version        | 16.2
 SDWebImage Version      | 5.14.2
 Integration Method      | Swift Package Manager
 Xcode Version           | Xcode 14.2
 Repro rate              | all the time (100%)
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Hi, We found a performance issue with ioQueue in SDWebImage.
The specific steps to reproduce and the reason for the problem are as follows:
1. First, load the image using `SDWebImageManager.shared.loadImage`, set `imageTransformer` or `imageThumbnailPixelSize` (whichever one you use, or both, will reproduce the problem), and cache the processed image in disk by setting `storeCacheType` to `.all`.
```swift
SDWebImageManager.shared.loadImage(
    with: url,
    options: [],
    context: [
        .imageThumbnailPixelSize: /* whatever you want */,
        .imageTransformer: /* whatever you want */,
        .storeCacheType: SDImageCacheType.all.rawValue
    ],
    completed: { ... }
)
```
2. SDWebImage will download the image and apply the transformer and thumbnail we set.
3. after that, since we set storeCacheType = .all, SDWebImage will cache the processed images on disk. We traced this logic and found that SDWebImage will encode the processed image in the ioQueue when caching.
<img width="1256" alt="image" src="https://user-images.githubusercontent.com/1300501/208222866-7c8547bf-8ba8-4eb2-b077-dc86cf07e69e.png">
4. Here is where the problem occurs. Some images are slow to encode (especially in WebP format, which takes seconds to encode), and ioQueue is a serial queue, which means that all the tasks that follow in ioQueue are blocked and waiting for the encoding to finish.
<img width="917" alt="image" src="https://user-images.githubusercontent.com/1300501/208225581-656b398f-48ae-4c23-a48b-23479e801bcf.png">
That is, after that, all cache queries, cache stores are blocked. If you call SDWebImageManager.shared.loadImage, even if the image is already cached in disk, it may take a long time to return the result because it needs to wait for previous encoding task in the ioQueue to finish.

In other words, if an ImageView in a TableViewCell or CollectionViewCell sets the image in this way, it will remain blank for a while after the list is reloaded or the cell is reused.

**Comment by dreampiggy** [comment]
The `SDImageCache` level IOQueue is to use gurantee IO-Safe, not Thread-safe.

Assume when you remove / store / query the same key **at the same time**. (multiple thread envionment)

However, this can be a option for advanced user to customize, or a better solution. Currently brutally put all Cache Operation (query/remove/store) is not a good idea on performance, just because it's simple.

**Comment by dreampiggy** [comment]
You can reply for some better idea about this, how we can solve this or provide a better customized behavior ?

1. Use key to distinguish

Create a Map<String, Queue>, the same key to query/remove/store has to wait previous op to finish. Different keys does not wait each other.

We can actually drop DispatchQueue and write a self-created queue manager instead of this.

2. Just ignore queue and let it go.

This may cause some IO-unsafe, like query a removing key will return success (should be fail). But I think this can still be acceptable. We can take care to use `NSData.writeTo:atomically:` to make sure atomic behavior and not result a broken image files.

**Comment by dreampiggy** [comment]
By the way...For that WebP images.

Here history a issue talking because of libwebp's encoding is slow on old iPhone.

We introdce that `cacheSerializer`, did you get it what's that for ?

https://github.com/SDWebImage/SDWebImage/pull/2245

**Comment by dreampiggy** [comment]
Decide to provide a new API on SDImageCacheConfig:

```objc
/**
 * The dispatch queue attr for ioQueue. You can config the QoS and concurrent/serial to internal IO queue. The ioQueue is used by SDImageCache to access read/write for disk data.
 * Defaults we use `DISPATCH_QUEUE_SERIAL`(NULL), to use serial dispatch queue to ensure single access for disk data. It's safe but may be slow.
 * @warning For advanced user only. Use when you meet IO performance issue during large disk cache query/write.
 * @note You can override this to use `DISPATCH_QUEUE_CONCURRENT`, use concurrent queue. **MAKE SURE** to keep `diskCacheWritingOptions` to use `NSDataWritingAtomic`, or this may cause corrupted disk data (because multiple thread write single disk cache without atomic is not IO-safe).
 * @note This value does not support dynamic changes. Which means further modification on this value after cache initialized has no effect.
 */
@property (strong, nonatomic, nullable) dispatch_queue_attr_t ioQueueAttributes;
```

**Comment by dreampiggy** [comment]
For Swift user, you need a bridge header to call that `DISPATCH_QUEUE_CONCURRENT` macro.

Like this:

```c
// Bridge.h
extern dispatch_queue_attr_t swift_dispatch_queue_concurrent(void);
// Bridge.m
dispatch_queue_attr_t swift_dispatch_queue_concurrent(void) {
    return DISPATCH_QUEUE_CONCURRENT
}
// AppDelegate.swift
SDImageCacheConfig.default.ioQueueAttributes = swift_dispatch_queue_concurrent()
```

**Comment by Madimo** [comment]
Hi, thanks for the help.
We tried using cacheSerializer to convert WebP images to PNG and it does perform better. it doesn't solve all the problems but it does work much better and we will use this solution for now.
For what you mentioned later, providing an ioQueueAttributes to change the ioQueue to become a concurrent queue, will this cause resource racing issue? Does SDWebImage internally use NSDataWritingAtomic for all cache accesses?

**Comment by dreampiggy** [comment]
Defaults to `NSDataWritingAtomic` (See `diskCacheWritingOptions`), which will write the file to temp folder firstly and then replace the target path with temp path.

If you use concurrent io queue, DO NOT Change it into other value.

**Comment by dreampiggy** [comment]
@Madimo Another changes in https://github.com/SDWebImage/SDWebImage/pull/3466

Previously we have a option `SDWebImageWaitStoreCache`, I previously think it should works (if you don't provide this option and call `[SDImageCache diskImageDataForKey:]` should return empty)

But however, since we `dispatch all IO related thing into the serial IO queue`, actually even you  don't provide this option, we still have to waiting for Image Encoding finish (consume more times)

In #3466, I seperate the `Encode Queue` and `IO related queue` into different thing. Now this option works as expected and you have to add this `SDWebImageWaitStoreCcahe` option if you call query inside completionBlock

**Comment by dreampiggy** [comment]
I will put this in 5.15.0's changelog and explain the old behavior.
